docs/tensor-ir.rst

```diff
@@ -0,0 +1,253 @@
+    void-type                   = "void"
+
+Scalar types
+============
+
+.. code:: abnf
+
+    scalar-type                 = int-type / floating-type
+    int-type                    = ("i" / "u") ("8" / "16" / "32" / "64")
+    floating-type               = "f" ("32" / "64")
+
+Scalar types are either integer ("i"), unsigned integer ("u"),
+or floating point ("f").
+The number behind the scalar type prefix denotes the number of bits,
+e.g. "f64" are double precision floating point numbers.
+
+Memref type
+===========
+
+.. code:: abnf
+
+    memref-type                 = "memref<" matrix-shape "," leading-dimension, "," offset ">"
+    matrix-shape                = scalar-type 2("x" integer-constant)
+    leading-dimension           = integer-constant
+    offset                      = integer-constant
+
+A memref points to a region of memory that stores a matrix.
+The underlying scalar type and the matrix shape is given by the ``matrix-shape`` rule.
+The leading dimension is the distance in number of scalars between rows.
+
+.. admonition:: TODO
+
+    - Memref should be extended to arbitrary order tensors
+    - Clarify offset
+    - Do we need a value type?
+
+Batch type
+==========
+
+.. code:: abnf
+
+    batch-type                  = "batch<" (scalar-type / memref-type) "," batch-layout ">"
+    batch-layout                = strided-layout / pointers-layout
+    strided-layout              = "strided<" integer-constant ">"
+    pointers-layout             = "pointers"
+
+The batch type describes a batch of scalars or memrefs.
+The batch is either given in a single memory region with a fixed
+distance between items (strided layout) or a pointer to each item is given (pointers layout).
+
+Instructions
+============
+
+.. code:: abnf
+
+    block                       = *instruction
+    instruction                 = value-instruction / axpy-instruction / barrier-instruction / lifetime-stop-instruction / matmul-instruction
+    value-instruction           = local-identifier "=" (alloca-instruction / get-work-item-instruction / submatrix-instruction)
+
+Alloca
+------
+
+.. code:: abnf
+
+    alloca-instruction          = "alloca" memref-type
+
+Overview
+........
+
+The alloca instruction allocates temporary memory that is freed automatically.
+
+Arguments
+.........
+
+The argument is the type of the returned value.
+
+Get work item
+-------------
+
+.. code:: abnf
+
+    get-work-item-instruction   = "get_work_item" local-identifier ["," local-identifier]
+
+Overview
+........
+
+Get work item fetches an item from a batch.
+
+Arguments
+.........
+
+The first operand must have the batch type.
+The optional second operand must be an integer scalar type and is used to specify
+an offset.
+
+Submatrix
+---------
+
+.. code:: abnf
+
+    submatrix-instruction       = "submatrix" local-identifier "[" slice "," slice "]"
+    slice                       = integer-constant ":" integer-constant
```

Comment:
What about adding a stride here? (or is that something for a higher-level IR instead?)
Probably, that would be something if `memref` will only support 2D tensors (i.e. matrices). Otherwise one could just reshape the tensor instead and take what's useful.